### PR TITLE
[tests] add traffic analysis for Cert_5_2_04_REEDUpgrade.py

### PR DESCRIPTION
**Pull request by hjian2017**

### CODE REVIEW

File: tests/scripts/thread-cert/Cert_5_2_04_REEDUpgrade.py [hunk 1]
```diff
@@ -82,14 +97,80 @@ def test(self):
             time.sleep(5)
             self.assertEqual(self.nodes[i].get_state(), 'router')
 
-        self.nodes[REED].start()
+        # 2 DUT_REED: Attach to any node, 2-hops from the Leader.
+        self.nodes[DUT_REED].start()
         time.sleep(5)
-        self.assertEqual(self.nodes[REED].get_state(), 'child')
+        time.sleep(ROUTER_SELECTION_JITTER)
```
Comment (bukepo) [inline tests/scripts/thread-cert/Cert_5_2_04_REEDUpgrade.py]: Is it possible to sleep shorter? Maybe by setting router selection jitter in `setUp`?

File: tests/scripts/thread-cert/Cert_5_2_04_REEDUpgrade.py [hunk 2]
```diff
@@ -82,14 +97,80 @@ def test(self):
...
+        reed_messages = self.sniffer.get_messages_sent_by(DUT_REED)
+
+        # The DUT_REED must not send a coap message here.
+        msg = reed_messages.does_not_contain_coap_message()
+        assert msg is True, "Error: The REED sent an Address Solicit Request"
+
+        # 3 DUT_REED: Sends a MLE Advertisement.
+        msg = reed_messages.next_mle_message(mle.CommandType.ADVERTISEMENT)
+        msg.assertSentWithHopLimit(255)
+        msg.assertSentToDestinationAddress("ff02::1")
```
Comment (bukepo) [inline tests/scripts/thread-cert/Cert_5_2_04_REEDUpgrade.py]: quote not consistent with this file.

File: tests/scripts/thread-cert/Cert_5_2_04_REEDUpgrade.py [hunk 3]
```diff
@@ -82,14 +97,80 @@ def test(self):
...
+        # 2 DUT_REED: Attach to any node, 2-hops from the Leader.
+        self.nodes[DUT_REED].start()
         time.sleep(5)
-        self.assertEqual(self.nodes[REED].get_state(), 'child')
+        time.sleep(ROUTER_SELECTION_JITTER)
+
+        reed_messages = self.sniffer.get_messages_sent_by(DUT_REED)
+
+        # The DUT_REED must not send a coap message here.
+        msg = reed_messages.does_not_contain_coap_message()
```
[the remark this answers is not in the document]
Comment (hjian2017) [reply tests/scripts/thread-cert/Cert_5_2_04_REEDUpgrade.py]: Yes， I think so. But you can see the same name in many scripts. I don't know if I should modify it in this script or not.

File: tests/scripts/thread-cert/Cert_5_2_04_REEDUpgrade.py [hunk 4]
```diff
@@ -82,14 +97,80 @@ def test(self):
...
+        # 3 DUT_REED: Sends a MLE Advertisement.
+        msg = reed_messages.next_mle_message(mle.CommandType.ADVERTISEMENT)
+        msg.assertSentWithHopLimit(255)
+        msg.assertSentToDestinationAddress("ff02::1")
+        msg.assertMleMessageContainsTlv(mle.SourceAddress)
+        msg.assertMleMessageContainsTlv(mle.LeaderData)
+        msg.assertMleMessageDoesNotContainTlv(mle.Route64)
+
+        # 4 Wait for REED_ADVERTISEMENT_INTERVAL+ REED_ADVERTISEMENT_MAX_JITTER seconds.
+        time.sleep(REED_ADVERTISEMENT_INTERVAL + REED_ADVERTISEMENT_MAX_JITTER)
```
[the remark this answers is not in the document]
Comment (hjian2017) [reply tests/scripts/thread-cert/Cert_5_2_04_REEDUpgrade.py]: Maybe we should put some longer running case out, let them run in parallel to save time

File: tests/scripts/thread-cert/Cert_5_2_04_REEDUpgrade.py (hunk 1 above)
Comment (hjian2017) [reply tests/scripts/thread-cert/Cert_5_2_04_REEDUpgrade.py]: “Maybe by setting router selection jitter in setUp?”
？？
I wasn't setting jitter.  Here means wait for ROUTER_SELECTION_JITTER seconds and then I can test 3rd step.

Comment (hjian2017) [reply tests/scripts/thread-cert/Cert_5_2_04_REEDUpgrade.py]: fixed

File: tests/scripts/thread-cert/Cert_5_2_04_REEDUpgrade.py [hunk 5]
```diff
@@ -82,14 +99,80 @@ def test(self):
             time.sleep(5)
             self.assertEqual(self.nodes[i].get_state(), 'router')
 
-        self.nodes[REED].start()
+        # 2 DUT_REED: Attach to any node, 2-hops from the Leader.
```
[the remark this answers is not in the document]
Comment (hjian2017) [reply tests/scripts/thread-cert/Cert_5_2_04_REEDUpgrade.py]: fixed.

File: tests/scripts/thread-cert/Cert_5_2_04_REEDUpgrade.py [hunk 6]
```diff
@@ -82,14 +99,79 @@ def test(self):
...
+
+        # 3 DUT_REED: Check if DUT_REED sent a correct MLE Advertisement or not.
+        msg = reed_messages.next_mle_message(mle.CommandType.ADVERTISEMENT)
+        msg.assertSentWithHopLimit(255)
+        msg.assertSentToDestinationAddress('ff02::1')
+        msg.assertMleMessageContainsTlv(mle.SourceAddress)
+        msg.assertMleMessageContainsTlv(mle.LeaderData)
+        msg.assertMleMessageDoesNotContainTlv(mle.Route64)
+
+        # 4 Wait for DUT_REED to send the second package.
```
Comment (jwhui) [inline tests/scripts/thread-cert/Cert_5_2_04_REEDUpgrade.py]: nit: package -> packet

File: tests/scripts/thread-cert/Cert_5_2_04_REEDUpgrade.py [hunk 7]
```diff
@@ -82,14 +99,79 @@ def test(self):
...
+        msg.assertMleMessageContainsTlv(mle.LeaderData)
+        msg.assertMleMessageContainsTlv(mle.LinkLayerFrameCounter)
+        msg.assertMleMessageContainsOptionalTlv(mle.MleFrameCounter)
+        msg.assertMleMessageContainsTlv(mle.Response)
+        msg.assertMleMessageContainsTlv(mle.Challenge)
+        msg.assertMleMessageContainsTlv(mle.LinkMargin)
+        msg.assertMleMessageContainsTlv(mle.Connectivity)
+        msg.assertMleMessageContainsTlv(mle.Version)
+
+        # 8 DUT_REED: Verify that its Address Solicit Request is properly.
```
Comment (jwhui) [inline tests/scripts/thread-cert/Cert_5_2_04_REEDUpgrade.py]: suggestion: you can simplify to "8 DUT_REED: Verify Address Solicit Request"

You can simplify the other comments in the same way.

**Comment by jwhui** [comment]
@librasungirl, I'm fine with keeping the timeouts as they are for now.  We can look to reducing the Travis times in a separate PR.

File: tests/scripts/thread-cert/Cert_5_2_04_REEDUpgrade.py (hunk 6 above)
Comment (hjian2017) [reply tests/scripts/thread-cert/Cert_5_2_04_REEDUpgrade.py]: fixed.

File: tests/scripts/thread-cert/Cert_5_2_04_REEDUpgrade.py (hunk 7 above)
Comment (hjian2017) [reply tests/scripts/thread-cert/Cert_5_2_04_REEDUpgrade.py]: fixed:)

**Comment by LuDuda** [comment]
@jwhui Have you considered to speed up regression tests by changing `only internal implementation` of alarm (in posix platform) to set up timer to value lower than given e.g. to time_ms/10 (clock base 100us), only in some TEST mode? Which may significantly speed up travis checks.

**Pull request merged**